mfr/server/handlers/export.py

```diff
@@ -37,61 +37,49 @@ class ExportHandler(core.BaseHandler):
         else:
             cache_file_path_str = '/export/{}'.format(self.cache_file_id)
-        self.cache_file_path = await self.cache_provider.validate_path(cache_file_path_str)
-
-        self.source_file_path = await self.local_cache_provider.validate_path(
-            '/export/{}'.format(self.source_file_id)
-        )
-
-        self.output_file_id = '{}.{}'.format(self.source_file_path.name, self.format)
-        self.output_file_path = await self.local_cache_provider.validate_path(
-            '/export/{}'.format(self.output_file_id)
-        )
-        self.metrics.merge({
-            'output_file': {
-                'id': self.output_file_id,
-                'path': str(self.output_file_path),
-                'provider': self.local_cache_provider.NAME,
-            }
-        })
 
     async def get(self):
-        """Export a file to the format specified via the associated extension library"""
+        """Export a file to the format specified via the associated extension
+        library.
 
+        First, try to export the file directly - this is only valid if the file is
+        already in the correct format.
+
+        Next, if caching is enabled, try to use a cached version.
+
+        Finally, do the actual conversion and export the converted file.
+        """
+
+        self._set_headers()
+
+        # File is already in the requested format
+        if self.metadata.ext == ".{}".format(self.format):
+            await self.write_stream(self.provider.download())
+            logger.info('Exported with no conversion.')
+            self.metrics.add('export.conversion', 'noop')
+            return
+
+        # Try to get a cached version
         if settings.CACHE_ENABLED:
             try:
-                cached_stream = await self.cache_provider.download(self.cache_file_path)
+                self.cache_file_path = await self.cache_provider.validate_path(cache_file_path_str)
+                await self.write_stream(self.cache_provider.download(self.cache_file_path))
+                logger.info('Cached file found; Sending downstream [{}]'.format(self.cache_file_path))
+                self.metrics.add('cache_file.result', 'hit')
+                return
+
+            # Cache miss
             except DownloadError as e:
                 assert e.code == 404, 'Non-404 DownloadError {!r}'.format(e)
                 logger.info('No cached file found; Starting export [{}]'.format(self.cache_file_path))
                 self.metrics.add('cache_file.result', 'miss')
-            else:
-                logger.info('Cached file found; Sending downstream [{}]'.format(self.cache_file_path))
-                self.metrics.add('cache_file.result', 'hit')
-                self._set_headers()
-                return await self.write_stream(cached_stream)
 
-        await self.local_cache_provider.upload(
+        # File isn't cached and it needs to be converted
+        await self.write_stream(await utils.bind_convert(
```

Comment:
While this is technically true, adding support for the noop also required some changes to `prepare`  because some objects were no longer necessary. Making that logic work inside of the exporter rather than the handler turned out to be a little simpler to write.

Reply:
It's become neccesary because we require the id inside the exporter  now.